rust/relay/ebpf-turn-router/src/move_payload.rs

```diff
@@ -0,0 +1,245 @@
+}
+
+/// Single entry point to add/remove bytes at the start of the UDP payload using only adjust_tail.
+/// - `delta < 0`: add `-delta` bytes (extend tail, slide payload toward tail, optionally write header)
+/// - `delta > 0`: remove `delta` bytes (slide payload toward head, then shrink tail)
+#[inline(always)]
+pub fn move_payload<const IP_HEADER_LEN: usize, const DELTA: i32>(
+    ctx: &XdpContext,
+    header: Option<&CdHdr>,
+) -> Result<(), Error> {
+    if DELTA < 0 {
+        slide_backward_and_extend::<IP_HEADER_LEN, DELTA>(ctx, header)
+    } else {
+        slide_forward_and_shrink::<IP_HEADER_LEN, DELTA>(ctx)
+    }
+}
+
+/* -------------------- Implementation Helpers -------------------- */
+
+#[inline(always)]
+fn payload_offset<const IP_HEADER_LEN: usize>() -> usize {
+    EthHdr::LEN + IP_HEADER_LEN + UdpHdr::LEN
+}
+
+/// Small, stack-safe copy: N must be <= 64 (kept out-of-line to avoid stack blow-up).
+#[inline(never)]
+fn copy_n<const N: usize>(ctx: &XdpContext, src: usize, dst: usize) -> Result<(), Error> {
+    let mut buf = [0u8; N];
+    let p = buf.as_mut_ptr() as *mut c_void;
+
+    let r1 = unsafe { bpf_xdp_load_bytes(ctx.ctx, src as u32, p, N as u32) };
+    if r1 < 0 {
+        return Err(Error::XdpLoadBytesFailed(r1));
+    }
+    let r2 = unsafe { bpf_xdp_store_bytes(ctx.ctx, dst as u32, p, N as u32) };
+    if r2 < 0 {
+        return Err(Error::XdpStoreBytesFailed(r2));
+    }
+    Ok(())
+}
+
+/// Copy fixed 64-byte chunks (out-of-line to avoid stack blow-up).
+#[inline(never)]
+fn copy_64(ctx: &XdpContext, src: usize, dst: usize) -> Result<(), Error> {
+    copy_n::<64>(ctx, src, dst)
+}
+
+/// Copy any remaining bytes after 64-byte chunks using greedy fixed sizes.
+#[inline(never)]
+fn copy_remainder(
+    ctx: &XdpContext,
+    src: usize,
+    dst: usize,
+    n: usize,
+) -> Result<(), Error> {
+    let mut src = src;
+    let mut dst = dst;
+    let mut remaining = n;
+    
+    if remaining >= 32 {
+        copy_n::<32>(ctx, src, dst)?;
+        src += 32;
+        dst += 32;
+        remaining -= 32;
+    }
+    if remaining >= 16 {
+        copy_n::<16>(ctx, src, dst)?;
+        src += 16;
+        dst += 16;
+        remaining -= 16;
+    }
+    if remaining >= 8 {
+        copy_n::<8>(ctx, src, dst)?;
+        src += 8;
+        dst += 8;
+        remaining -= 8;
+    }
+    if remaining >= 4 {
+        copy_n::<4>(ctx, src, dst)?;
+        src += 4;
+        dst += 4;
+        remaining -= 4;
+    }
+    if remaining >= 2 {
+        copy_n::<2>(ctx, src, dst)?;
+        src += 2;
+        dst += 2;
+        remaining -= 2;
+    }
+    if remaining >= 1 {
+        copy_n::<1>(ctx, src, dst)?;
+    }
+    Ok(())
+}
+
+/* -------------------- ADD path (delta < 0) -------------------- */
+
+/// Extend tail by `add`, slide payload toward tail by `add`, and optionally write `header`.
+#[inline(never)]
+fn slide_backward_and_extend<const IP_HEADER_LEN: usize, const DELTA: i32>(
```

Comment:
Given that we always use delta has 4 or -4 and have two different functions now, should we just get rid of that parameter?

`slide_backward_and_extend` is always called with -4.